### ISSUE TITLE
lizardfs HA (URAFT +VPS) [question]

### ISSUE THREAD

**Issue by eleaner** [lizardfs/lizardfs#848]
Hi,

I am trying to understand how lizardfs works in HA configuration and how could I use it on my three VPS cluster.

I understand that uraft will try to assign the defined floating IP to one of the three masters.
something very similar to how keepalived works.

Does it mean that the setup is pretty much dedicated for local network and will likely not work with the VPS? 
Mainly because I am not able to assign IP in the datacentre local IP nd public IP probably only in some and only invoking additional API.

**Comment by zicklag** [comment]
I had the same question. It seems like it is currently impossible to use uRaft on public clouds ( or at least AWS as far as I have seen ). The is an AWS blog post that says:

> All traffic is unicast (the Address Resolution Protocol, or ARP, is gone too). The VPC forwarding design knows where all IP addresses are, and proxies ARP responses for all ARP requests locally at the hypervisor.

I'm not super network savvy, but I'm pretty sure that means that LizardFS can't use ARP broadcasting to create a floating IP on AWS.

I contacted the team through email about the issue and they said that they would get back to me when they can, so we'll see what they come up with.

**Comment by Zorlin** [comment]
Yep, in most public clouds you cannot use the uRaft solution.

There should be some ways around it - some people have developed failover solutions that use the shadow masters + manually moving around a floating IP to provide HA.

**Comment by Zorlin** [comment]
If you can figure out the rest of the glue (starting/stopping/moving around masters etc etc) here's an example of doing something like this in Pacemaker/Corosync - https://medium.com/@gt.anand1994/ha-cluster-with-elasticip-using-corosync-and-pacemaker-a013d288ae8

**Comment by eleaner** [comment]
Thank you @zicklag @Zorlin 
I am also not too much network savvy but just wonder if HA solution (not uRaft of course) could work by simply specifying multiple network addresses.
let's say there are three masters M1 M2 and M3
each master knows all the other IP
each client knows All the IP
the algorithm magic makes it all connect where required and makes sure that there is only one acting master.

does it sound like a reasonable feature request?

I was also thinking if in such a setup the read operations to master could be load balanced between the three. The shadows should be able to serve them quite successfully.

**Comment by Zorlin** [comment]
I believe that's roughly how MooseFS PRO works.

**Comment by eleaner** [comment]
@Zorlin I am not sure. 
As I understand the clients of Moose accesses a single hostname that points to all IP addresses. So a default round-robin failover will eventually connect to the acting master.
I am talking about a more sophisticated approach

@pbeza does it make sense to raise this as a feature request?

**Comment by zicklag** [comment]
[SeaweedFS](https://github.com/chrislusf/seaweedfs) has you specify multiple IP addresses and I ( just a guess ) think it might just try them each until one works.

I read a bit about [TiKV](https://github.com/tikv/tikv) in their blog and, if I remember right, clients can connect to any of the masters and whichever master the client hits will forward the requests to the active one. If a client can't connect to one of the masters, then it will just try another one.

I really think it would be good for the people who have to operate and deploy LizardFS to have LizardFS manage the HA without depending on specialized network configuration. It seems like it would be relatively simple to implement, but I don't know enough about it to make a good estimate.

**Comment by Zorlin** [comment]
@eleaner no, in the PRO version of MooseFS I believe the clients know about all masters. It uses DNS to discover them, but does not rely on failing until it succeeds... it just gets a list of all of them and then talks to them.

**Comment by pbeza** [comment]
> does it sound like a reasonable feature request?

Yep – it's totally reasonable feature request.

Currently LizardFS uRaft uses ARP broadcasting to create a floating IP assigned to the active LizardFS master. ARP broadcasting cannot be easily used on cloud and currently there is no easy workaround.

**Comment by eleaner** [comment]
@pbeza 
I checked the seaweedfs and I can confirm that HA-wise it is precisely what I am thinking about
each master knows other masters, masters decide who is the leader among each other
for clients is enough to know only one master - communication is redirected accordingly (but I guess it is more HA if clients know all the masters too).

regarding temp workaround, as mentioned above I run three masters on three VPS.
One is always acting master and the other two are shadows.
lizardfs elements communicate through zerotier network. The acting master in addition to his own IP receives MasterIP that simulates floating IP (using zerotier API)
in case the master VPS goes down - one of the shadows takes over and MasterIP is reassigned.
I use a docker swarm to control the switching but all lizardfs containers use host network for communication.

known issues
1. zerotier is not the fastest medium, I could use wireguard but I did not find out how to simulate floating IP there
2. my setup cannot at the moment identify that acting master crashed. it will switch only when VPS is down. 
2. new VPS to act as the master is selected randomly without the knowledge of lizardfs state. The shadow could be down, stuck or behind yet still will be selected.

I think a temporary solution could be the ability to override liardfs action that changes floating IP. I could then plugin the zerotier API there.

**Comment by zicklag** [comment]
I had an idea earlier about a possible workaround. It might require the development of a small proxy utility, but it wouldn't require changes to LizardFS:

Using uRaft, the LizardFS masters already know how to elect a leader to handle the requests. The uRaft package comes with a CLI that can show you what the current acting master is. The trick is then to run a proxy service on each of the masters that takes in requests from the chunkservers and clients and forwards them to the active master.

Because the LizardFS clients and chunkservers are not setup to allow providing multiple master addresses you would also have to run HAProxy or an equivalent tool on each of the chunkservers and clients that would listen on localhost and proxy the requests to any one of the available masters. Each chunkserver would have the master set to localhost on the proxy's port.
